src/main/python/docs/source/guide/python_end_to_end_tut.rst

```diff
@@ -0,0 +1,561 @@
+"dummycode", we transform this one column into 3 different columns, one for each bin.
+
+At last, we make use of the "recode" transformation for categorical columns, it maps all distinct categories in
+the column into consecutive numbers, starting from 1. In our example we recode the "income" column, which
+transforms it from "<=$50K" and ">$50K" to "1" and "2" respectively.
+
+Another good resource for further ways of processing is: https://apache.github.io/systemds/site/dml-language-reference.html
+
+There we provide different examples for defining jspec's and what functionality is currently supported.
+
+After defining the .jspec file we can read it by passing the filepath, data_type and value_type using the following command:
+
+.. code-block:: python
+
+    dataset_jspec = "adult/jspec.json"
+    jspec = sds.read(dataset_jspec, data_type="scalar", value_type="string")
+
+Finally, we need to define a custom dml file to split the features from the labels and replace certain values, which we will need later.
+We will call this file "preprocess.dml":
+
+.. code-block::
+
+    get_X = function(matrix[double] X, int start, int stop)
+    return (matrix[double] returnVal) {
+    returnVal = X[start:stop,1:ncol(X)-1]
+    }
+    get_Y = function(matrix[double] X, int start, int stop)
+    return (matrix[double] returnVal) {
+    returnVal = X[start:stop,ncol(X):ncol(X)]
+    }
+    replace_value = function(matrix[double] X, double pattern , double replacement)
+    return (matrix[double] returnVal) {
+    returnVal = replace(target=X, pattern=pattern, replacement=replacement)
+    }
+
+The get_X function simply extracts every column except the last one and can also be used to pick certain slices from the dataset.
+The get_Y function only extracts the last column, which in our case holds the labels. Replace_value is used to replace a double value with another double.
+The preprocess.dml file can be read with the following command:
+
+.. code-block:: python
+
+    preprocess_src_path = "preprocess.dml"
+    PREPROCESS_package = sds.source(preprocess_src_path, "preprocess", print_imported_methods=True)
+
+The print_imported_methods flag can be used to verify whether every method has been parsed correctly.
+
+Step 3: Applying the preprocessing steps
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Generally speaking, we would use the transform_encode function on the train dataset and with the returned encoding call the transform_apply function on the test dataset.
+In the case of the Adult dataset, we have inconsistent label names inside the test dataset and the train dataset, which is why we will show how we can deal with that using SystemDS.
+First of all, we combine the train and the test dataset by using the rbind() function. This function simply appends the Frame F2 at the end of Frame F1.
+This is necessary to ensure that the encoding is identical between train and test dataset.
```

Comment:
the encoding should be extracted from the training data, and then applied to the test data, this is why we have two different outputs from transform_encode. you can also argue that you need to save the M, to be able to produce predictionson new data, since you need to apply the same preprocessing on unseen data.

if i remember correctly the code is something like:

X, M = X.transform_encode(spec=jspec)
X_test = X_test.transform_apply(M)